### PR TITLE
Fix up cookie handling.

### LINKED ISSUE #497: Errors with cookie handling during decompression.
Spotted by @hawkowl.

RFC 7540 Section 8.1.2.5 reads as follows:

> To allow for better compression efficiency, the Cookie header field MAY be split into separate header fields, each with one or more cookie-pairs.  If there are multiple Cookie header fields after decompression, these MUST be concatenated into a single octet string using the two-octet delimiter of 0x3B, 0x20 (the ASCII string "; ") before being passed into a non-HTTP/2 context, such as an HTTP/1.1 connection, or a generic HTTP server application.

We are very much in violation of that MUST right now. This is breaking some Twisted applications that work with multiple cookies, whereby cookies are getting lost because Twisted (rightly) expects only a single cookie header.

For the moment, this issue tracks only the MUST. Later on we may want to tackle the other side of this (the MAY).

My belief is that the last sentence of that paragraph makes this a hyper-h2 issue, not a hpack issue: HTTP/2 contexts may want to see those headers individually, but in general we shouldn't show them to users that way. This means we need to add another step to the inbound header normalization pipeline that grabs all cookie fields and concatenates them together.

### PR DESCRIPTION
**Pull request by Lukasa**
Resolves #497.

This is a breaking change, and I'd describe it as a moderately urgent one, so I'm hereby pronouncing that our next release will be 3.0. I'll be opening issues shortly to handle removing all deprecated functionality that needs to be stripped.

### CODE REVIEW

File: h2/stream.py [hunk 1]
```diff
@@ -1316,6 +1309,30 @@ def _build_headers_frames(self,
         frames[-1].flags.add('END_HEADERS')
         return frames
 
+    def _process_received_headers(self,
+                                  headers,
+                                  header_validation_flags,
+                                  header_encoding):
+        """
+        When headers have been received from the remote peer, runs a processing
```
Comment (alexwlchan) [inline h2/stream.py]: nit: "run a processing pipeline"?

File: test/test_basic_logic.py [hunk 2]
```diff
@@ -850,6 +850,85 @@ def test_changing_max_frame_size(self, frame_factory, frame_size):
...
+
+        cookie_field_count = sum(1 for n, _ in e.headers if n == 'cookie')
+        assert cookie_field_count == 1
+
+        for n, v in e.headers:
+            if n == 'cookie':
+                assert v == expected
+                break
+        else:
+            pytest.mark.fail("Unable to check cookie header")
```
Comment (alexwlchan) [inline test/test_basic_logic.py]: When would we ever come down the `else:` branch of this loop?

If we pass the assert above, we know that there’s exactly one header whose name is `cookie`. So unless `e.headers` changes in the second iteration, we’ll always run into the `break` statement.

How about something like this?

```python
cookie_fields = [(n, v) for (n, v) in e.headers if n == 'cookie']
assert len(cookie_fields) == 1

_, v = cookie_fields[0]
assert v == expected
```

File: test/test_basic_logic.py [hunk 3]
```diff
@@ -850,6 +850,85 @@ def test_changing_max_frame_size(self, frame_factory, frame_size):
...
+        events = c.receive_data(f.serialize())
+
+        assert len(events) == 1
+        e = events[0]
+
+        cookie_field_count = sum(1 for n, _ in e.headers if n == 'cookie')
+        assert cookie_field_count == 3
+
+        received_cookies = [(n, v) for n, v in e.headers if n == 'cookie']
+        assert cookie_headers == received_cookies
```
Comment (alexwlchan) [inline test/test_basic_logic.py]: Could we tidy this up in a similar way?

```python
received_cookies = [(n, v) for n, v in e.headers if n == 'cookie']

assert len(received_cookies) == 3
assert cookie_headers == received_cookies
```

File: h2/stream.py [hunk 4]
```diff
@@ -1105,14 +1102,10 @@ def receive_headers(self, headers, end_stream, header_encoding):
...
-            headers = validate_headers(headers, hdr_validation_flags)
-
-        if header_encoding:
-            headers = list(_decode_headers(headers, header_encoding))
-
-        events[0].headers = headers
+        hdr_validation_flags = self._build_hdr_validation_flags(events)
+        events[0].headers = self._process_received_headers(
+            headers, hdr_validation_flags, header_encoding
+        )
```
Comment (alexwlchan) [inline h2/stream.py]: 👍

**Comment by Lukasa** [comment]
Thanks for that feedback @alexwlchan! I think I've addressed it in the last two commits.

**Comment by Lukasa** [comment]
Thanks @alexwlchan! :heart:

**Pull request merged**